File: Day 3/solve.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jan 15 22:37:18 2021

@author: lando
"""

import numpy as np

def map_wire(li):
    coor = [0,0]
    wire = []
    for segment in li:
        axis = int(segment[0] in ('U','D'))
        dist = (-1 if segment[0] in ('D','L') else 1) * int(segment[1:])
        wire.append({'coor':tuple(coor),'axis':axis,'dist':dist})
        coor[axis] += dist
    return wire

def is_intersection(seg1,seg2):
    coor = [0,0]
    if seg1['axis'] == seg2['axis']:
        return None
    else:
        axis1 = seg1['axis']
        before1 = np.sign(seg1['coor'][axis1]-seg2['coor'][axis1])
        after1 = np.sign(seg1['coor'][axis1]+seg1['dist']-seg2['coor'][axis1])
        if before1 == after1:
            return None
        axis2 = seg2['axis']
        before2 = np.sign(seg2['coor'][axis2]-seg1['coor'][axis2])
        after2 = np.sign(seg2['coor'][axis2]+seg2['dist']-seg1['coor'][axis2])
        if before2 == after2:
            return None
        coor[axis2] = seg1['coor'][axis2]
        coor[axis1] = seg2['coor'][axis1]
        return tuple(coor)

# Part 1
lines = [line.strip().split(',') for line in open('C:/Users/lando/Desktop/Python/Advent of Code 2019/Day 3/input.txt','r').readlines()]
wires = [map_wire(line) for line in lines]
intersections = []
for seg1 in wires[0]:
    for seg2 in wires[1]:
        x = is_intersection(seg1,seg2)
        if x is not None: intersections.append(x)
        
ans = 9999999999
for x in intersections:
    ans = min(ans,abs(x[0])+abs(x[1]))
print(ans)

# Part 2
def delay(wire, coor):
    dist = 0
    for seg in wire:
        axis = seg['axis']
        if seg['coor'][1-axis] != coor[1-axis]:
            dist += abs(seg['dist'])
        elif np.sign(seg['coor'][axis]-coor[axis]) != np.sign(seg['coor'][axis]+seg['dist']-coor[axis]):
            dist += abs(seg['coor'][axis]-coor[axis])
            return dist
        else:
            dist += abs(seg['dist'])

ans = 999999999
for x in intersections:
    print(x)
    ans = min(ans,delay(wires[0],x)+delay(wires[1],x))
print(ans)

File: Day 12/Part 2/solve.jl
function step(p::Array{Int64,1},v::Array{Int64,1})
    for i in 1:4
        v[i] += sum([(p[i] < p[j]) for j in 1:4])
        v[i] -= sum([(p[i] > p[j]) for j in 1:4])
    end
    p += v
    return p, v
end

function period(p::Array{Int64,1},v::Array{Int64,1})
    periods = Int64[0, 0, 0, 0]
    done = Int8[0, 0, 0, 0]
    p_0, v_0 = copy(p), copy(v)
    n = 0
    while true
        p, v = step(p, v)
        n += 1
        done = ((p == p_0) && (v == v_0))
        if done
            return n
        end
    end
end

function gcd(x::Int,y::Int)
    return x/numerator(x//y)
end

function solve(p::Array{Int64,2},v::Array{Int64,2})
    periods = Int64[]
    for i in 1:3
        push!(periods, period(p[:,i],v[:,i]))
    end
    return reduce((x,y) -> convert(Int,x*y/gcd(x,y)),periods)
end

filepath = "C:/Users/lando/Desktop/Python/Advent of Code 2019/Day 12/Part 2/input.txt"
p = open(filepath) do file
    lines = readlines(file)
    p = Int64[]
    for line in lines
        line = replace(replace(line,">" => ""),"<" => "")
        vars = split(line,", ")
        for var in vars
            eval(Meta.parse(var))
        end
        if length(p) == 0
            p = [x y z]
        else
            p = [p; x y z]
        end
    end
    return p
end

v = [0 0 0
     0 0 0
     0 0 0
     0 0 0]

answer = solve(p,v)
println(answer)